include/deal.II/differentiation/ad/ad_number_traits.h

```diff
@@ -0,0 +1,1043 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2016 - 2017 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE at
+// the top level of the deal.II distribution.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_differentiation_ad_ad_number_traits_h
+#define dealii_differentiation_ad_ad_number_traits_h
+
+#include <deal.II/base/numbers.h>
+#include <deal.II/base/exceptions.h>
+#include <deal.II/differentiation/ad/ad_number_types.h>
+
+#include <boost/type_traits.hpp>
+
+#include <complex>
+#include <type_traits>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Differentiation
+{
+  namespace AD
```

Comment:
this namespace is not documented, it seems

Reply:
This namespace will be documented in a unifying header, which I did not include in this PR. I suppose I can just commit the salient parts of that file.